### PR TITLE
Add new types of gates, fix rotation angle, add docstrings

### PR DESCRIPTION
**Pull request by vsergeyev**

### CODE REVIEW

**Comment by vsergeyev** [comment]
* Please note, there are conversions for XX, YY, ZZ gates in `append_gate()` but it is no such names in https://github.com/rigetti/pyquil/blob/master/pyquil/gates.py
 * There are `TOFFOLI` and `FREDKIN` gates in Qulacs, but no methods to  directly add them in qulacs circuit. It  may be worth to submit a ticket in Qulacs repo.

File: src/python/qequlacs/utils.py [hunk 1]
```diff
@@ -1,5 +1,15 @@
...
+ * corresponding Qulacs gates appended to circuit
+ * if no matching gate on Qulacs side - unitary matrix used
+ * for > 3 qubits exception is raised (not implemented yet)
+"""
+
 import qulacs
-from zquantum.core.circuit import Circuit, Gate
 import numpy as np
 
 def convert_circuit_to_qulacs(circuit):
```
Comment (simonwa7) [inline src/python/qequlacs/utils.py:15]: I'm going to be a stickler here and ask for proper docstrings on all the functions in this file.

File: src/python/qequlacs/utils.py [hunk 2]
```diff
@@ -11,8 +21,23 @@ def convert_circuit_to_qulacs(circuit):
...
+    # From z-quantum.core _circuit.py: exponent=gate.params[0] / np.pi
+    # So gate.params[0] = exponent * np.pi
+    if gate.params:
+        angle = gate.params[0]
+    else:
+        angle = 0
+
+    modifiers = [] # dagger and so on
+    if hasattr(gate,  'modifiers'):
+        modifiers = gate.modifiers
```
Comment (simonwa7) [inline src/python/qequlacs/utils.py:39]: I'm assuming these are "Zapata" gates in which case I don't believe we use modifiers currently

File: src/python/qequlacs/utils.py [hunk 3]
```diff
@@ -38,24 +73,45 @@ def append_gate(gate, qulacs_circuit):
...
             qulacs_circuit.add_SWAP_gate(index_1, index_2)
         elif gate.name == 'XX':
-            qulacs_circuit.add_multi_Pauli_rotation_gate([index_1, index_2], [1,1], -gate.params[0]*2/np.pi)
+            qulacs_circuit.add_multi_Pauli_rotation_gate([index_1, index_2], [1,1], -angle)
         elif gate.name == 'YY':
-            qulacs_circuit.add_multi_Pauli_rotation_gate([index_1, index_2], [2,2], -gate.params[0]*2/np.pi)
+            qulacs_circuit.add_multi_Pauli_rotation_gate([index_1, index_2], [2,2], -angle)
         elif gate.name == 'ZZ':
-            qulacs_circuit.add_multi_Pauli_rotation_gate([index_1, index_2], [3,3], -gate.params[0]*2/np.pi)
+            qulacs_circuit.add_multi_Pauli_rotation_gate([index_1, index_2], [3,3], -angle)
```
Comment (simonwa7) [inline src/python/qequlacs/utils.py:86]: Can someone explain why we're dropping the factor of `2/np.pi`?

File: src/python/qequlacs/utils.py [hunk 4]
```diff
@@ -38,24 +73,45 @@ def append_gate(gate, qulacs_circuit):
...
         elif gate.name == 'CPHASE':
-            mat = np.diag([1., np.exp(1.j*np.pi*gate.params[0])])
-            gate = qulacs.gate.DenseMatrix(index_2, mat)
-            gate.add_control_qubit(index_1,1)
+            matrix = np.diag([1., np.exp(1.j * angle)])
```
Comment (simonwa7) [inline src/python/qequlacs/utils.py:88]: Again, why are we dropping the additional factor of `np.pi`?

**Review by simonwa7 — CHANGES_REQUESTED** [review]

File: src/python/qequlacs/utils.py (hunk 3 above)
Comment (mstechly) [reply src/python/qequlacs/utils.py:86]: Different frameworks use gate parameters with different conventions – with/without pi, etc.
And @vsergeyev figured out that these multipliers here are a correct way to do that.

Since I had issues with these before, I wouldn't merge it until it gives consistent results with other frameworks. If it doesn't then we'll fix it.

File: src/python/qequlacs/utils.py (hunk 1 above)
Comment (vsergeyev) [reply src/python/qequlacs/utils.py:15]: agree

File: src/python/qequlacs/utils.py (hunk 2 above)
Comment (vsergeyev) [reply src/python/qequlacs/utils.py:39]: Right, these are gates from z-quantum-core,
Modifiers allow to get Sdag, Tdag gates.
If it is not planned to use - we may remove this.

File: src/python/qequlacs/utils.py (hunk 3 above)
Comment (vsergeyev) [reply src/python/qequlacs/utils.py:86]: @simonwa7 this is from the assumption we have 
`exponent=gate.params[0]/np.pi`
in https://github.com/zapatacomputing/z-quantum-core/blob/master/src/python/zquantum/core/circuit/_circuit.py#L757

and in qulacs_simulator it is:
`qulacs_circuit.add_multi_Pauli_rotation_gate(indices, [1, 1], -np.pi * op.gate._exponent)`

so this result in:

`qulacs_circuit.add_multi_Pauli_rotation_gate(indices, [1, 1], -np.pi * gate.params[0] / np.pi)`

File: src/python/qequlacs/utils.py (hunk 4 above)
Comment (vsergeyev) [reply src/python/qequlacs/utils.py:88]: @simonwa7  please see my logical path in an answer above,
hope it is a right way

File: src/python/qequlacs/simulator.py [hunk 5]
```diff
@@ -43,6 +49,9 @@ def run_circuit_and_measure(self, circuit, **kwargs):
         return Measurements(bitstrings)
 
     def get_expectation_values(self, circuit, qubit_operator, **kwargs):
+        """TODO: Needs a good explanation what is  it and how
```
Comment (vsergeyev) [inline src/python/qequlacs/simulator.py:52]: @mstechly here I really need to get better understanding

File: src/python/qequlacs/simulator_test.py [hunk 6]
```diff
@@ -1,9 +1,149 @@
...
+        self.assertEqual(qulacs_circuit.get_gate(2).get_name(), "Pauli-rotation")
+
+    def test_gate_ZZ(self):
+        def_gate, gate = ZZ(0, 0, 1)
+        self.assertEqual(gate.name, "ZZ")
+        self.assertEqual(len(gate.qubits), 2)
+        self.assertEqual(len(gate.params), 1)
+        self.assertEqual(def_gate.name, "ZZ")    
+
+    def test_append_circuit_YY(self):
```
Comment (mstechly) [inline src/python/qequlacs/simulator_test.py:143]: There's a typo here, should be `ZZ`.

File: src/python/qequlacs/simulator_test.py [hunk 7]
```diff
@@ -1,9 +1,149 @@
...
+        self.assertEqual(len(gate.params), 1)
+        self.assertEqual(def_gate.name, "XX")
+
+    def test_append_circuit_XX(self):
+        # Given
+        circuit = Circuit(Program(H(0), CNOT(0, 1), XX(0, 0, 1)))
+        # When
+        qulacs_circuit = convert_circuit_to_qulacs(circuit)
+        self.assertEqual(qulacs_circuit.get_gate_count(), 3)
+        self.assertEqual(qulacs_circuit.get_gate(2).get_name(), "Pauli-rotation")
```
Comment (mstechly) [inline src/python/qequlacs/simulator_test.py:119]: I think this test doesn't really check whether we're getting the correct result for this gate.
Also, testing it with angle equal to 0 doesn't really check whether we got the multipliers right.

File: src/python/qequlacs/gates.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+def XX(angle: ParameterDesignator, q1: QubitDesignator, q2: QubitDesignator) -> Gate:
+    """Produces a XX gate which is the tensor product of X with X such as XX=X⊗X::
+
+        XX(phi) = [[cos(phi / 2), 0, 0, -1j * sin(phi / 2)],
+                   [0, cos(phi / 2), -1j * sin(phi / 2), 0],
+                   [0, -1j * sin(phi / 2), cos(phi / 2), 0],
+                   [-1j * sin(phi / 2), 0, 0, cos(phi / 2)]]
+
+    :param angle: The angle to rotate around the x-axis on the bloch sphere.
```
Comment (mstechly) [inline src/python/qequlacs/gates.py:36]: We're using this other docstring convention with:
```
Args:
    q1: lorem ipsum
...
```

File: src/python/qequlacs/gates.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+from pyquil.quil import DefGate
+from pyquil.quilatom import Parameter, quil_sin, quil_cos, quil_exp
+from pyquil.quilbase import (
+    AbstractInstruction,
+    Gate
+)
+
+
+def XX(angle: ParameterDesignator, q1: QubitDesignator, q2: QubitDesignator) -> Gate:
```
Comment (mstechly) [inline src/python/qequlacs/gates.py:28]: @simonwa7 @dexter2206 – don't you think this should actually go to `z-quantum-core`?

File: src/python/qequlacs/utils.py [hunk 10]
```diff
@@ -38,31 +76,54 @@ def append_gate(gate, qulacs_circuit):
...
+        else:
+            unitary_matrix = gate.to_unitary()
+            qulacs_circuit.add_dense_matrix_gate([index_1, index_2, index_3], unitary_matrix)
     else:
         n_qubits = len(gate.qubits)
         Exception(str(n_qubits) + "- qubit gates not supported.")
     return qulacs_circuit
 
 
 def qubitop_to_qulacspauli(qubit_operator):
```
Comment (mstechly) [inline src/python/qequlacs/utils.py:124]: I've also noticed we lack typing in this module.
`qubit_operators` should be of type `openfermion.QubitOperator`

**Review by mstechly — CHANGES_REQUESTED** [review]

File: src/python/qequlacs/simulator.py (hunk 5 above)
Comment (mstechly) [reply src/python/qequlacs/simulator.py:52]: Looking for some good educational materials on expectation values.
Also, the difference between `get_expectation_values` and `get_exact_expectation_values` is that the latter one uses wavefunction, so an exact representation of the quantum state that's impossible to do on the real device, but saves you a lot of time if you use it with a simulator.

File: src/python/qequlacs/gates.py (hunk 9 above)
Comment (dexter2206) [reply src/python/qequlacs/gates.py:28]: https://github.com/zapatacomputing/z-quantum-core/blob/dd6fb59af2b7695cb6a8ad1fe62891e36d73e578/src/python/zquantum/core/circuit/gates/_two_qubit_gates.py#L43 I think it is in the new branch with circuits and gates rework :)

File: src/python/qequlacs/simulator_test.py (hunk 6 above)
Comment (dexter2206) [reply src/python/qequlacs/simulator_test.py:143]: Actually, it has a funny consequence: we have a repeated method name, so the latter overrides the former. Therefore, the first test method is not even visible when launching test-suite and is never run.

Btw. most IDEs highlight such a repeated method name as an error, and so do static analysis tools, e.g. pylint, so it might be worthwhile to integrate such a tool in our everyday workflows :slightly_smiling_face:

Comment (vsergeyev) [reply src/python/qequlacs/simulator_test.py:143]: +

File: src/python/qequlacs/simulator_test.py (hunk 7 above)
Comment (vsergeyev) [reply src/python/qequlacs/simulator_test.py:134]: added cases with Pi and Pi/2

File: src/python/qequlacs/gates.py (hunk 9 above)
Comment (mstechly) [reply src/python/qequlacs/gates.py]: @dexter2206 how far are you from merging it to `dev`?

File: src/python/qequlacs/simulator_test.py [hunk 11]
```diff
@@ -1,9 +1,178 @@
...
+        # Given
+        n_samples = 100
+        # XX with Pi will -> both qubits to 0 in ~100% cases
+        prog1 = Program(X(0), CNOT(0, 1), XX(np.pi, 0, 1))
+        # XX with Pi/2 will -> both qubits to 0 in ~50% cases
+        prog2 = Program(X(0), CNOT(0, 1), XX(np.pi/2, 0, 1))
+        circuit1 = Circuit(prog1)
+        circuit2 = Circuit(prog2)
+        # When
+        results1 = self.qvm.run_and_measure(prog1, trials=10)
```
Comment (mstechly) [inline src/python/qequlacs/simulator_test.py:124]: You define `n_samples=100` but the use `trials=10` – seems inconsistent.
Also, having 10 trials is not enough for testing, as tests might sometimes fail due to some edge cases – much better to set it to 100 or even 1000.

File: src/python/qequlacs/simulator.py [hunk 12]
```diff
@@ -25,6 +25,12 @@
 
 
 class QulacsSimulator(QuantumSimulator):
+    """Quantum simulator class to use as a backend in z-quantum-core.
```
Comment (mstechly) [inline src/python/qequlacs/simulator.py:28]: Wrong docstring format.

**Review by mstechly — COMMENTED** [review]
1. Qulacs code still lacks typing.
2. Given that @dexter2206 mentioned that XX gates are already implemented by him in pyQuil, I think it'll make sense to wait until his changes are on `dev` and then clean this part.

File: src/python/qequlacs/simulator_test.py (hunk 11 above)
Comment (vsergeyev) [reply src/python/qequlacs/simulator_test.py:124]: got it

File: src/python/qequlacs/simulator.py (hunk 12 above)
Comment (vsergeyev) [reply src/python/qequlacs/simulator.py:28]: fixed

**Comment by vsergeyev** [comment]
Fixed issues found during review, it should be a bit better right now.

**Comment by mstechly** [comment]
Fixed in #15 , closing!
